feat(canvas): allow removing planes

marstamm opened the pull request:
This enables us to remove planes, e.g. when they are renamed.

This is used in https://github.com/bpmn-io/bpmn-js/pull/1538. Example use case here:
https://github.com/bpmn-io/bpmn-js/blob/0370ae612341d2c20227bb6b20bfd9fb6f2669c9/lib/features/modeling/behavior/SubProcessPlaneBehavior.js#L155-L173
<!--

Thanks for creating this pull request!

Please make sure to link the issue you are closing as "Closes #issueNr". 
This helps us to understand the context of this PR.

-->

nikku commented on lib/core/Canvas.js:
```diff
@@ -401,11 +401,17 @@ Canvas.prototype.setActivePlane = function(plane) {
     svgAttr(this._activePlane.layer, 'display', 'none');
-  }
 
+    // svg should only have one associated element
```
Could you elaborate on this? Why are we doing what we're doing here?

marstamm replied:
We have a link from the svg to the currently active root element (`data-element-id="..."`), which we update on plane switch: https://github.com/bpmn-io/diagram-js/blob/91626d03b756573951c687b72ff21b81c0a3db80/lib/core/Canvas.js#L421

We also have a reference from the element to the svg, which we did not update on plane switches previously.

When we removed an (inactive) plane + root element, it would also remove the `data-element-id` from the svg, even when it was not the "active" element. This breaks mouse interaction with the diagram.

nikku replied:
Which exact _mouse interaction_? Just trying to understand!

marstamm replied:
Specifically scrolling the canvas. The events are captured here:
https://github.com/bpmn-io/diagram-js/blob/1063f7c18474096d3d7c9e400ce82a1bf762a157/lib/features/interaction-events/InteractionEvents.js#L82-L92

for the root, this would lead to 
 * `gfx === canvas._svg`
 * and because of `data-element-id === ""` ==> `element === null`

it would not find an element and not fire mouse events

nikku replied:
Got it. Makes sense (and is obviously a bug!).

nikku commented on lib/core/Canvas.js:
```diff
@@ -464,6 +470,30 @@ Canvas.prototype.findPlane = function(element) {
+  }
+
+  // remove root element
+  this.setRootElementForPlane(null, plane, true);
```
What if plane === activePlane? Any precautions we'd like to set?

nikku replied:
Also, this particular part needs further clarification. What is happening here (and why)?

marstamm replied:
I was thinking about it with the renaming use-case in mind. There, we would want to reuse the existing root element on the newly created plane, and therefore need to remove it first.

I think deleting a plane should work like [`canvas._clear()`](https://github.com/bpmn-io/diagram-js/blob/91626d03b756573951c687b72ff21b81c0a3db80/lib/core/Canvas.js#L229-L252) on a plane level. 
So in case it is the active plane, it should set the active plane to `null`. The integrating library takes care of setting either a new one, or it resets to the default plane. (this is still missing)

nikku replied:
You are raising an important point here: How is this supposed to be used (i.e. to rename vs. add and remove)? What responsibility lies within the Canvas and what within the integrating library?

If we got clarity on this, please consider adding an integration test to verify this particular scenario in a more complex setup (i.e. renaming a plane, with contents, as the `id` changes or whatever).

I'm fine to keep the API low-level, if it is clear how it is supposed to be used to accomplish certain tasks, i.e. renaming.

marstamm replied:
Agreed, I'll get on that :+1:

nikku commented on test/spec/core/CanvasSpec.js:
```diff
@@ -2337,6 +2337,65 @@ describe('Canvas', function() {
+        canvas.removePlane('b');
+
+        // then
+        expect(svgAttr(canvas._svg, 'data-element-id')).to.equal('A');
```
Please always (also) test via high-level APIs: `ElementRegistry.get(...)`.

This is closest to how we'll use it (and our users will use it, too).

nikku reviewed (commented):
Some questions, but generally looking good.

I'd love to better understand what actually broke so we added updating of the root element.

nikku commented:
Specifically regarding https://github.com/bpmn-io/bpmn-js/blob/0370ae612341d2c20227bb6b20bfd9fb6f2669c9/lib/features/modeling/behavior/SubProcessPlaneBehavior.js#L168 I'm wondering why we do not support _renaming_ of planes, as their identifiers change.

marstamm commented:
I was under the impression that a low-level API is enough and we can get by on delete+create.

However, I now see the problems with it, especially on complex diagrams. I'll add something similar to `elementRegistry.updateId` for planes as well

marstamm commented:
I modified the `removePlane` method to work more similar to a `_clear` and remove all children. 
I also added a renaming method, as my initial Idea of removing- and adding the plane again does not work well when all children are removed

nikku commented:
As discussed, let us ensure the following regarding the `removePlane` API:

* Be atomic in what it does (no internal recursion in removing children and so forth)
* Be forgiving (no magic assignment of, i.e. new active plane); we assume the caller will clean up the mess when inconsistent state is the result of an intermediate canvas operation (cf. make collaboration behavior in bpmn-js).

The `_clear` method is a rare exception as now editing or viewer operations take place after it has been invoked.

nikku reviewed (approved):
Thanks for following up. Works for me, also against `bpmn-js@master`.
